Improved push and pull with upload manager and download manager

aaronlehmann opened the pull request:
This commit adds a transfer manager which deduplicates and schedules
transfers, and also an upload manager and download manager that build on
top of the transfer manager to provide high-level interfaces for uploads
and downloads. The push and pull code is modified to use these building
blocks.

Some benefits of the changes:
- Simplification of push/pull code
- Pushes can upload layers concurrently
- Failed downloads and uploads are retried after backoff delays
- Cancellation is supported, but individual transfers will only be
  cancelled if all pushes or pulls using them are cancelled.
- The distribution code is decoupled from Docker Engine packages and API
  conventions (i.e. streamformatter), which will make it easier to split
  out.

This commit also includes unit tests for the new distribution/xfer
package. The tests cover 87.8% of the statements in the package.

fixes https://github.com/docker/docker/issues/18498

Commit:
Improved push and pull with upload manager and download manager

This commit adds a transfer manager which deduplicates and schedules
transfers, and also an upload manager and download manager that build on
top of the transfer manager to provide high-level interfaces for uploads
and downloads. The push and pull code is modified to use these building
blocks.

Some benefits of the changes:

- Simplification of push/pull code
- Pushes can upload layers concurrently
- Failed downloads and uploads are retried after backoff delays
- Cancellation is supported, but individual transfers will only be
  cancelled if all pushes or pulls using them are cancelled.
- The distribution code is decoupled from Docker Engine packages and API
  conventions (i.e. streamformatter), which will make it easier to split
  out.

This commit also includes unit tests for the new distribution/xfer
package. The tests cover 87.8% of the statements in the package.

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

LK4D4 commented:
It doesn't change design really. I think we can move to code-review.

jfrazelle commented:
agreed

LK4D4 commented on daemon/daemon.go:
```diff
@@ -731,7 +735,8 @@ func NewDaemon(config *Config, registryService *registry.Service) (daemon *Daemo
 	}
 
-	distributionPool := distribution.NewPool()
+	d.downloadManager = xfer.NewLayerDownloadManager(d.layerStore, 5)
```
Maybe 5 should be a const. Also why 5?

aaronlehmann replied:
Yes, this should probably be defined in a constant.

The number 5 is arbitrary and just seemed like a sane number of simultaneous transfers. It could be made tunable if people think it's worth the effort.

LK4D4 commented on daemon/daemon.go:
```diff
@@ -1032,23 +1036,56 @@ func (daemon *Daemon) TagImage(newTag reference.Named, imageName string, force b
+		if prog.Message != "" {
+			formatted := sf.FormatStatus(prog.ID, prog.Message)
+			if _, err := outStream.Write(formatted); err != nil {
+				cancelFunc()
```
Shouldn't you return from func here?

aaronlehmann replied:
The function needs to continue to consume messages from `progressChan` until it's closed, otherwise the writer could block forever.

aaronlehmann replied:
> The function needs to continue to consume messages from progressChan until it's closed, otherwise the writer could block forever.

To clarify this: cancellation is asynchronous, so there can be a race between cancelling a push or pull, and getting further progress messages.

dmcgowan replied:
Can we at least log the error here then

aaronlehmann replied:
Cancelling the context causes an error to be logged by the pull code:

    DEBU[1239] Error trying v2 registry: context canceled

If we logged an error here, it might end up in the log several times, because many progress messages could be in flight during the cancellation. So I think we would need to make sure to only log once.

dmcgowan replied:
Not as much concerned that something gets log rather than specifically that `err` is logged somewhere. What type of error is this usually and is there a specific error it is expecting here.

aaronlehmann replied:
Modified it to log err the first time write fails.

LK4D4 commented on daemon/daemon.go:
```diff
@@ -1032,23 +1036,56 @@ func (daemon *Daemon) TagImage(newTag reference.Named, imageName string, force b
 func (daemon *Daemon) PullImage(ref reference.Named, metaHeaders map[string][]string, authConfig *cliconfig.AuthConfig, outStream io.Writer) error {
+	// Include a buffer so that slow client connections don't affect
+	// transfer performance.
+	progressChan := make(chan xfer.Progress, 100)
```
why 100? :)

tonistiigi commented on daemon/daemon.go:
```diff
@@ -1032,23 +1036,56 @@ func (daemon *Daemon) TagImage(newTag reference.Named, imageName string, force b
 	}
 
-	return distribution.Pull(ref, imagePullConfig)
+	err := distribution.Pull(ctx, ref, imagePullConfig)
```
How about reusing the context for progress.

```
ctx, cancel := context.WithCancel(context.Background())

if !isTty {
	p, ctx = progress.New(ctx)

	go func() {
		err := p.WriteTo(outStream)
		// or p.Marshal(transformer).WriteTo(outStream)
		if err != nil {
			cancel()
		}
	}
}
// ^ this could even be in postImagesCreate

err := distribution.Pull(ctx, ref, imagePullConfig)
p.Wait() // or writesDone channel
return err


// in progress generator

p = progress.FromContext(ctx)
p.Write(&progress.Item{}) // normally would use helpers instead
```


One thing I don't like about current method is that if we discussed the pull interface cleanup(not in this PR) one idea was that `ImagePullConfig` would not be a place to dump random stuff, but only contain `MetaHeaders` and `AuthConfig` (things that client sends per pull). This heavily reuses the big config struct model again.

aaronlehmann replied:
I can see the appeal of doing this, but I'm reluctant to store objects on contexts in general. I don't like how it loses type safety and needs to be wrapped with special utility functions to do the proper type assertions. I also don't like the way variables are looked up by string names, adding overhead and making name conflicts possible. Contexts just remind me too much of Python :)

What do other people think?

`ImagePullConfig` definitely needs to be cleaned up. This could either happen as part of this PR, or as a followup. I'm thinking the daemon could create a `Distribution` object at startup that references static things like the image store and layer store. `Pull` and `Push` could be methods on that object.

tonistiigi replied:
I think the convention for name conflicts if to use package name prefix. https://github.com/golang/net/blob/master/trace/trace.go#L266

stevvooe replied:
If you use a type, with a zero-value, it will package qualify the key:

```go
type myKey int

// ... later, in function lang
v := ctx.Value(myKey{})
```

No one outside the package can access that value.

stevvooe replied:
As far as putting injections on the context, I don't have a good answer. Seems like it should be part of a "pull struct", if we have one.

tonistiigi commented on distribution/pull_v2.go:
```diff
@@ -98,91 +92,101 @@ func (p *v2Puller) pullV2Repository(ref reference.Named) (err error) {
+	}
 
-	di.broadcaster.Write(p.sf.FormatProgress(stringid.TruncateID(digestStr), "Verifying Checksum", nil))
+	progressChan <- xfer.Progress{ID: ld.ID(), Action: "Verifying Checksum"}
```
would be nice to do:

```
progress.FromContext(ctx).Printf("Verifying Checksum")
progress.Printf(ctx, "Verifying Checksum")
```

aaronlehmann commented:
I did some benchmarks comparing this PR with the existing code on master. The numbers were almost exactly the same. I pulled golang:latest repeatedly from the hub, and from a local registry.

Using the hub, the average pull time with this patch was 27.55 seconds, versus 27.74 with master.

Using a local registry, the average pull time with this patch was 20.21 seconds, versus 20.01 seconds with master.

In both cases, the runtime is dominated by registering the layers, which can't be parallelized. The patched version could be somewhat faster in some cases, because it prioritizes downloading the bottom layers so registration can start earlier. My internet connection is fast enough that this doesn't show up much in my benchmark. With a slower connection, I think it could make a big difference.

This patch wraps downloads and registrations with an `io.Pipe` to support cancellation (see `cancelReadCloser`). Fortunately, this doesn't seem to have a significant performance impact.

aaronlehmann commented:
After some experimentation, I've found that changing the download concurrency limit from 5 to 3 helps with slower links. I've made this change.

I used Network Link Conditioner to simulate a 40 mbps network connection. Pulling wordpress:latest with the latest version of the PR takes 47.06s, vs 52.16s on master. Pulling golang:latest takes 62.42s, vs 66.19s on master.

The differences here are caused by fewer simulatenous downloads allowing the base layers to complete first. Then they can be unpacked and registered while other layers are still downloading.

dmcgowan commented on distribution/pull_v2.go:
```diff
@@ -98,91 +92,101 @@ func (p *v2Puller) pullV2Repository(ref reference.Named) (err error) {
+			logrus.Errorf("Failed to remove temp file: %s", tmpFile.Name())
+		}
+
+		return nil, 0, err
```
Will this cause a retry?

aaronlehmann replied:
Yes. The thinking is that if we receive something that doesn't match the digest, it could have been corrupted in transit, for example by a misbehaving proxy, or a broken server in a load balancing rotation. Retrying has a chance of getting the expected content.

There is also an argument to be made that if we don't get the right data the first time, we aren't likely to get it after a retry either.

I'm not sure what makes most sense in this case.

aaronlehmann commented:
I've added another commit that cleans up the progress monitoring.

It unifies the channel-based ProgressReader under `distribution/xfer` with the traditional version in `pkg/progressreader`, creating `pkg/progress` which replaces both. It adds a method to `StreamFormatter` that creates a channel to pass to `NewProgressReader` for automatic stream formatting of progress items. I've updated existing `progressreader` users to use this, and also replaced `writeDistributionProgress`. Finally, there are new utility methods in `pkg/progress` to output simple or printf-formatted messages or updates to a progress channel.

If people feel like this is the right direction, I'll squash the commits before merge.

aaronlehmann commented:
Updated the new commit after some discussion with @tonistiigi. Now the common `ProgressReader` uses an interface to send updates instead of writing to a channel directly. This keeps it synchronous for the use cases that output directly to `StreamFormatter`. PTAL.

LK4D4 commented:
I think previous implementation of progress was cleaner. It was so nice to have progress without crappy streamformatter.

aaronlehmann commented:
Rebased and squashed commits.

LK4D4 commented on distribution/pull_v2.go:
```diff
@@ -98,98 +93,109 @@ func (p *v2Puller) pullV2Repository(ref reference.Named) (err error) {
-		return
+		logrus.Debugf("Error statting layer: %v", err)
+		if err == distribution.ErrBlobUnknown {
+			return nil, 0, xfer.DoNotRetry{err}
```
I wonder why go vet doesn't complain about unkeyed fields :/

aaronlehmann replied:
I've seen go vet do that sometimes when it can't import dependencies becuase they haven't been compiled yet: https://github.com/docker/distribution/pull/811

Annoying that it cares in this case. The struct only has one item. Maybe I should change the type from `struct` to `error` so it's not necessary to use keyed fields when initializing it.

aaronlehmann replied:
Actually, I don't think that would work, because then the type assertion would always succeed. So I'll just add keys to the initializers.

go vet also complains about this one: `pull_v1.go:39: registry.ErrNoSupport composite literal uses unkeyed fields`

LK4D4 replied:
Yeah, np. I just wonder why CI doesn't work properly.

LK4D4 replied:
@aaronlehmann Should we add `go install ./...` in Dockerfile for that?

aaronlehmann replied:
@LK4D4 Not sure it's that simple - you might need to set build tags to get the right things built. It will also add build time. Moving `validate-vet` after `binary` might fix it, depending how `binary` builds stuff.

LK4D4 replied:
I think it'll be different containers :/ We need to build binary in validate container too. Ok, thanks for help!

LK4D4 commented:
LGTM

LK4D4 commented:
@aaronlehmann seems like there is deadlock in tests.

aaronlehmann commented:
@LK4D4 fixed

thaJeztah commented:
@aaronlehmann needs a rebase now :cry:

aaronlehmann commented:
@thaJeztah rebased

thaJeztah commented:
ping @LK4D4 @tonistiigi PTAL

(Thanks for the rebase @aaronlehmann)

LK4D4 commented:
still LGTM for me

aaronlehmann commented:
I noticed `TestSuccessfulDownload` failed on experimental. This was due to a race in the test itself. It was checking the final progress output before releasing the download (which flushes progress). I've fixed this issue.

dmcgowan commented on distribution/xfer/download.go:
```diff
@@ -0,0 +1,420 @@
+	}()
+
+selectLoop:
+	for {
```
This for loop is not necessary since both conditions of the select break or return

stevvooe commented:
@aaronlehmann Any numbers for improvements on push performance?

dmcgowan commented:
LGTM

aaronlehmann commented:
@stevvooe I used Network Link Conditioner to simulate a 40mbps link with 100 ms latency. Pushing redis to the hub took 182 seconds on master, and 75 seconds with this patch.

Even without a high-latency link, the patched version is a lot better: with Network Link Conditioner turned off, I get 60s on master and 22s with this patch for the same push. S3 latency on the hub really hurts when the push process is completely serial.

The differences would be even more pronounced with an image that has many small layers.

stevvooe commented:
@aaronlehmann Excellent.

I wonder how this will affect the adoption of parallel gzip (https://github.com/docker/docker/pull/17674).

stevvooe commented:
@csfrancis @bobrik Would you mind seeing if this affects push/pull times positively? Parallelization layer push should see much better CPU utilization for compression.

stevvooe commented on distribution/xfer/transfer.go:
```diff
@@ -0,0 +1,314 @@
+	}
+
+	masterProgressChan := make(chan progress.Progress)
+	xfer := xferFunc(masterProgressChan, start, inactive)
```
_G(s)_

stevvooe commented:
LGTM

aaronlehmann commented:
I did some code review with @tonistiigi and he pointed out the argument to `makeDownloadFunc` wasn't generated from `initialRootFS`, which could cause problems with Windows base layers. I've repushed with a fix for this. It actually makes the code slightly simpler. Tonis is about to test this PR on Windows.

I'm comfortable with the PR as it is now, assuming the Windows testing doesn't uncover problems. In the reviews, a few ideas for improvements have come up, especially replacing `ImagePullConfig` / `ImagePushConfig` with something saner. These changes should probably go in a separate PR.

aaronlehmann commented:
Thanks @tonistiigi for testing on Windows. We found and fixed an issue with the way `RootFS` objects were constructed in the presence of a Windows base layer. Push and pull work on Windows with the latest revision of the PR.

csfrancis commented:
Unfortunately, I don't really have time to work on this right now. I'll try to get a colleague to look into this.

tonistiigi commented:
LGTM

Does this require changes to docs as it changes what happens when a pull is cancelled?

thaJeztah commented:
@tonistiigi, possibly, but not sure we explicitly describe the behavior in the docs. I'll have a quick look

thaJeztah commented:
Looks like there's a mention of "pull cancelation" in https://github.com/docker/docker/blob/master/docs/reference/commandline/build.md, but I couldn't find other references really. If we want to describe the behavior, I'd suggest adding it to the push/pull docs.

aaronlehmann commented:
This reference is `Note: Currently only the "run" phase of the build can be canceled until pull cancellation is implemented).` I think removing it should be sufficient.

Personally, I don't think it's necessary to mention that killing a `docker pull` or `docker push` process will cancel a push or pull, because that's the behavior that one would expect. But I'm happy to add it if people think there's value to stating it explicitly. Thoughts?

tonistiigi commented:
I agree that it is what people should expect, but I think we should define it because we are changing it(probably even with "Starting from Docker v1.10..."). I think it's even more important for the API side where currently someone may be using the API without actually keeping the connection open for the duration of the pull.

Commit:
Update docs for addition of transfer manager

Closing the HTTP connection requesting a push or pull will cancel the
push or pull. This behavior also applies to the CLI.

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

aaronlehmann commented:
I've updated the docs in a new commit. PTAL.

I decided not to include language like "Starting from Docker v1.10...". For the CLI, it's not especially important to know that this behavior changed, and it's even a little embarrassing to mention that cancellation wasn't supported until recently. API docs are already versioned, so it doesn't seem necessary to mention change between versions in-band.

thaJeztah commented:
Some minor nits, LGTM otherwise 

ping @moxiegirl @SvenDowideit for review

aaronlehmann commented:
@thaJeztah: Changes made.

cpuguy83 commented on distribution/xfer/download.go:
```diff
@@ -0,0 +1,424 @@
+				close(progressChan)
+			}()
+
+			select {
```
Isn't this select equivelant to:
```go
<-start
```

aaronlehmann replied:
Yes, it's based on `makeDownloadFunc`, which outputs progress in the default case. I'll simplify this one.

thaJeztah commented:
Looks like windows failure is because of https://github.com/docker/docker/pull/18397, which I thought was fixed in CI?

colrack commented:
I tested this.
In my case pushing images (> 1GB and mutiple layers) to a private v2 registry it's faster now. As far as I can see this is due to the fact that before this only 1 cpu core was used. On a single core machine the time taken is nearly the same; however in a multiple core cpu I can see all core are working at 100% while pushing (is this ok?).

aaronlehmann commented:
@colrack: Parallel compression helps, but another big effect comes from removing pauses between sequential registry operations. Your private registry may not have high latency, but registries backed by object stores typically do. With the old one-upload-at-a-time model, there could be big pauses between actual data transfers because of the latency on the registry side.

colrack commented:
I want to be more precise; my experience now is push speed improved x2 x3 x4 depending on the number of cores of the cpu and layers of the image. The bottleneck now is the fact 1 layer is pinned to 1 core; can this be improved? Why is the compression so heavy?
You're correct there is not high latency between registry and the push machine; however my registry is backed by azure blob storage.

aaronlehmann commented:
@colrack: Yes, there is discussion about compression being parallelized over multiple cores. See https://github.com/docker/docker/pull/17674.

cpuguy83 commented:
was moved to docs review while I was reviewing, but code LGTM  anyway :)

thaJeztah commented:
thanks! docs LGTM

ping@moxiegirl @SvenDowideit for review

moxiegirl commented:
I hear this is expedited work.  I'll LGTM this and create a carry PR.

LK4D4 commented:
@moxiegirl thanks

dts commented:
This feature has completely broken docker for my use case - I can no longer upload from my home computer, as large layers compete for limited bandwidth and time each other out.  Is there a way to disable this behavior at the command line?

stevvooe commented:
@dts Without more information, it is unlikely that this PR is the root cause of your particular issue without completing an analysis. Typically, reporting issues on merged PRs is an ineffective venue for solving your problem.

I'd recommend submitting another issue or reaching out the mailing list for help. Submit the issue and reference this as a possible cause. However, please ensure that you separate the description of the problem from the perceived cause, as it may be another cause.

dts commented:
@stevvooe, thanks for your patience and guidance - I've added an issue here: https://github.com/docker/docker/issues/20936.

stevvooe commented:
@dts No problem. Make sure to fill in that issue with as many details as possible. At least, it should include the docker version and the exact output when the error is encountered. From there, I'd recommend including logs from daemon in debug mode, as that will provide clues as to how many simultaneous streams are starting and their actual scope.

dts commented:
@stevvoe - It's more of a feature request for me - I'd like to reduce the concurrency of sending images, does having my docker version and output text help with that?

stevvooe commented:
@dts Its preferable to understand the root cause of the issue rather than adding another feature that may or may not address your problem. This PR is not really a feature and fixes to its behavior would be a feature, either.

dts commented:
@stevvooe makes sense - context is added to the fresh new issue, thanks again for your help!
